xarray/named_array/core.py

```diff
@@ -0,0 +1,311 @@
+        if len(dims) != self.ndim:
+            raise ValueError(
+                f"dimensions {dims} must have the same length as the "
+                f"number of data dimensions, ndim={self.ndim}"
+            )
+        return dims
+
+    @property
+    def attrs(self) -> dict[typing.Any, typing.Any]:
+        """Dictionary of local attributes on this variable."""
+        if self._attrs is None:
+            self._attrs = {}
+        return self._attrs
+
+    @attrs.setter
+    def attrs(self, value: typing.Mapping[typing.Any, typing.Any]) -> None:
+        self._attrs = dict(value)
+
+    @property
+    def data(self) -> typing.Any:
+        """
+        The Variable's data as an array. The underlying array type
+        (e.g. dask, sparse, pint) is preserved.
+
+        See Also
+        --------
+        Variable.to_numpy
+        Variable.as_numpy
+        Variable.values
+        """
+
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        if data.shape != self.shape:
+            raise ValueError(
+                f"replacement data must match the Variable's shape. "
+                f"replacement data has shape {data.shape}; Variable has shape {self.shape}"
+            )
+        self._data = data
+
+    @property
+    def real(self):
+        """
+        The real part of the variable.
+
+        See Also
+        --------
+        numpy.ndarray.real
+        """
+        return self._replace(data=self.data.real)
+
+    @property
+    def imag(self):
+        """
+        The imaginary part of the variable.
+
+        See Also
+        --------
+        numpy.ndarray.imag
+        """
+        return self._replace(data=self.data.imag)
+
+    def __dask_tokenize__(self):
+        # Use v.data, instead of v._data, in order to cope with the wrappers
+        # around NetCDF and the like
+        from dask.base import normalize_token
+
+        return normalize_token((type(self), self._dims, self.data, self._attrs))
+
+    def __dask_graph__(self):
+        return self._data.__dask_graph__() if is_duck_dask_array(self._data) else None
+
+    def __dask_keys__(self):
+        return self._data.__dask_keys__()
+
+    def __dask_layers__(self):
+        return self._data.__dask_layers__()
+
+    @property
+    def __dask_optimize__(self):
+        return self._data.__dask_optimize__
+
+    @property
+    def __dask_scheduler__(self):
+        return self._data.__dask_scheduler__
+
+    def __dask_postcompute__(self):
+        array_func, array_args = self._data.__dask_postcompute__()
+        return self._dask_finalize, (array_func,) + array_args
+
+    def __dask_postpersist__(self):
+        array_func, array_args = self._data.__dask_postpersist__()
+        return self._dask_finalize, (array_func,) + array_args
+
+    def _dask_finalize(self, results, array_func, *args, **kwargs):
+        data = array_func(results, *args, **kwargs)
+        return NamedArray(self._dims, data, attrs=self._attrs)
```

Comment:
```suggestion
        return type(self)(self._dims, data, attrs=self._attrs)
```
For subclassed namedarrays.